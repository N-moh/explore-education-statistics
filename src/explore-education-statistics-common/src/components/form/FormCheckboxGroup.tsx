import { Omit } from '@common/types/util';
import classNames from 'classnames';
import React, { Component, createRef } from 'react';
import FormCheckbox, {
  CheckboxChangeEventHandler,
  FormCheckboxProps,
} from './FormCheckbox';
import FormFieldset, { FieldSetProps } from './FormFieldset';

export type CheckboxOption = Omit<
  FormCheckboxProps,
  'name' | 'checked' | 'onChange'
>;

export type FormCheckboxGroupProps = {
  name: string;
  onAllChange?: CheckboxChangeEventHandler;
  onChange?: CheckboxChangeEventHandler;
  options: CheckboxOption[];
  selectAll?: boolean;
  small?: boolean;
  value: string[];
} & FieldSetProps;

interface State {
  checkedCount: number;
}

/**
 * Basic checkbox group that should be used as a controlled component.
 * When using Formik, use {@see FormFieldRadioGroup} instead.
 */
class FormCheckboxGroup extends Component<FormCheckboxGroupProps, State> {
  public static defaultProps = {
    legendSize: 'm',
    selectAll: false,
    small: false,
    value: [],
  };

  private ref = createRef<HTMLDivElement>();

  public componentDidMount(): void {
    if (this.ref.current) {
      import('govuk-frontend/components/checkboxes/checkboxes').then(
        ({ default: GovUkCheckboxes }) => {
          new GovUkCheckboxes(this.ref.current).init();
        },
      );
    }
  }

  public render() {
    const {
      value,
      onAllChange,
      onChange,
      name,
      id,
      options,
      selectAll,
      small,
    } = this.props;
    const isAllChecked = options.every(
      option => value.indexOf(option.value) > -1,
    );

    return (
      <FormFieldset {...this.props}>
        <div
          className={classNames('govuk-checkboxes', {
            'govuk-checkboxes--small': small,
          })}
          ref={this.ref}
        >
          {options.length > 1 && selectAll && (
            <FormCheckbox
              id={`${id}-all`}
              label="Select all"
              name={name}
              value="select-all"
              checked={isAllChecked}
              onChange={event => {
                if (onAllChange) {
                  onAllChange(event);
                }
              }}
            />
          )}

          {options.map(option => (
            <FormCheckbox
              {...option}
              name={name}
              key={option.id}
              checked={value.indexOf(option.value) > -1}
              onChange={event => {
                if (onChange) {
                  onChange(event);
                }
              }}
            />
          ))}
        </div>
      </FormFieldset>
    );
  }
}

export default FormCheckboxGroup;
