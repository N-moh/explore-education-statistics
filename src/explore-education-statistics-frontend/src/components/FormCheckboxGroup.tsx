import Checkboxes from 'govuk-frontend/components/checkboxes/checkboxes';
import React, { Component, createRef } from 'react';
import FormCheckbox, { CheckboxChangeEventHandler } from './FormCheckbox';
import FormFieldSet, { FieldSetProps } from './FormFieldSet';

export interface CheckboxOption {
  checked?: boolean;
  hint?: string;
  id: string;
  label: string;
  value: string;
}

type Props = {
  checkedValues: {
    [value: string]: boolean;
  };
  name: string;
  onAllChange?: CheckboxChangeEventHandler;
  onChange?: CheckboxChangeEventHandler<any>;
  options: CheckboxOption[];
} & Partial<FieldSetProps>;

interface State {
  checkedCount: number;
}

class FormCheckboxGroup extends Component<Props, State> {
  public static defaultProps: Partial<Props> = {
    checkedValues: {},
    legendSize: 'm',
  };

  private ref = createRef<HTMLDivElement>();

  public componentDidMount(): void {
    new Checkboxes(this.ref.current).init();
  }

  private handleChange: CheckboxChangeEventHandler = event => {
    if (this.props.onChange) {
      this.props.onChange(event);
    }
  };

  private handleAllChange: CheckboxChangeEventHandler = event => {
    if (this.props.onAllChange) {
      this.props.onAllChange(event);
    }
  };

  private renderCheckboxes() {
    const { checkedValues, onAllChange, name, options } = this.props;

    const checkedCount = Object.values(checkedValues).filter(Boolean).length;

    return (
      <div className="govuk-checkboxes" ref={this.ref}>
        {onAllChange && (
          <FormCheckbox
            id={`${name}-all`}
            label="Select all"
            name={name}
            value="select-all"
            checked={checkedCount === options.length}
            onChange={this.handleAllChange}
          />
        )}

        {options.map(option => (
          <FormCheckbox
            {...option}
            name={name}
            key={option.id}
            onChange={this.handleChange}
            checked={Boolean(checkedValues[option.value])}
          />
        ))}
      </div>
    );
  }

  public render() {
    const { legend, ...restProps } = this.props;

    return legend ? (
      <FormFieldSet {...restProps} legend={legend}>
        {this.renderCheckboxes()}
      </FormFieldSet>
    ) : (
      this.renderCheckboxes()
    );
  }
}

export default FormCheckboxGroup;
