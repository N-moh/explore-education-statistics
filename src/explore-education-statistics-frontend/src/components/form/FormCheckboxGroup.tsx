import Checkboxes from 'govuk-frontend/components/checkboxes/checkboxes';
import React, { Component, createRef } from 'react';
import FormCheckbox, { CheckboxChangeEventHandler } from './FormCheckbox';
import FormFieldSet, { FieldSetProps } from './FormFieldSet';

export interface CheckboxOption {
  hint?: string;
  id: string;
  label: string;
  value: string;
}

type Props = {
  name: string;
  onAllChange?: CheckboxChangeEventHandler;
  onChange?: CheckboxChangeEventHandler<any>;
  options: CheckboxOption[];
  value: string[];
} & FieldSetProps;

interface State {
  checkedCount: number;
}

class FormCheckboxGroup extends Component<Props, State> {
  public static defaultProps: Partial<Props> = {
    legendSize: 'm',
    value: [],
  };

  private ref = createRef<HTMLDivElement>();

  public componentDidMount(): void {
    new Checkboxes(this.ref.current).init();
  }

  public render() {
    const { value, onAllChange, name, options } = this.props;

    const checkedCount = Object.values(value).filter(Boolean).length;

    return (
      <FormFieldSet {...this.props}>
        <div className="govuk-checkboxes" ref={this.ref}>
          {onAllChange && (
            <FormCheckbox
              id={`${name}-all`}
              label="Select all"
              name={name}
              value="select-all"
              checked={checkedCount === options.length}
              onChange={event => {
                if (this.props.onAllChange) {
                  this.props.onAllChange(event);
                }
              }}
            />
          )}

          {options.map(option => (
            <FormCheckbox
              {...option}
              name={name}
              key={option.id}
              checked={value.indexOf(option.value) > -1}
              onChange={event => {
                if (this.props.onChange) {
                  this.props.onChange(event);
                }
              }}
            />
          ))}
        </div>
      </FormFieldSet>
    );
  }
}

export default FormCheckboxGroup;
