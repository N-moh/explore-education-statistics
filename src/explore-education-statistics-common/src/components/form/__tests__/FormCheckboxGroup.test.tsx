import React from 'react';
import { render } from 'react-testing-library';
import FormCheckboxGroup from '../FormCheckboxGroup';

describe('FormCheckboxGroup', () => {
  test('renders list of checkboxes in correct order', () => {
    const { container, getAllByLabelText } = render(
      <FormCheckboxGroup
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    const checkboxes = getAllByLabelText(/Test checkbox/);

    expect(checkboxes).toHaveLength(3);
    expect(checkboxes[0]).toHaveAttribute('value', '1');
    expect(checkboxes[1]).toHaveAttribute('value', '2');
    expect(checkboxes[2]).toHaveAttribute('value', '3');

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders checkboxes with some pre-checked', () => {
    const { container, getAllByLabelText } = render(
      <FormCheckboxGroup
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
        value={['2']}
      />,
    );

    const checkboxes = getAllByLabelText(/Test checkbox/) as HTMLInputElement[];

    expect(checkboxes).toHaveLength(3);
    expect(checkboxes[0].checked).toBe(false);
    expect(checkboxes[1].checked).toBe(true);
    expect(checkboxes[2].checked).toBe(false);

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with legend', () => {
    const { container, getByText } = render(
      <FormCheckboxGroup
        legend="Choose some checkboxes"
        id="test-checkboxes"
        name="test-checkboxes"
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    expect(getByText('Choose some checkboxes')).toBeDefined();
    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders correctly with small size variants', () => {
    const { container } = render(
      <FormCheckboxGroup
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        small
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    expect(container.querySelector('.govuk-checkboxes--small')).not.toBeNull();
    expect(container).toMatchSnapshot();
  });

  test('renders unchecked `Select all` option when `selectAll` is true', () => {
    const { container, getByLabelText } = render(
      <FormCheckboxGroup
        value={[]}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    const selectAllCheckbox = getByLabelText('Select all') as HTMLInputElement;

    expect(selectAllCheckbox.checked).toBe(false);

    expect(container.innerHTML).toMatchSnapshot();
  });

  test('renders checked `Select all` checkbox when all options are pre-checked', () => {
    const { getByLabelText } = render(
      <FormCheckboxGroup
        value={['1', '2', '3']}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    const selectAllCheckbox = getByLabelText('Select all') as HTMLInputElement;

    expect(selectAllCheckbox.checked).toBe(true);
  });

  test('renders `Select all` with small variant', () => {
    const { container } = render(
      <FormCheckboxGroup
        value={[]}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        small
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    expect(container.querySelector('.govuk-checkboxes--small')).not.toBeNull();
  });

  test('does not render checked `Select all` checkbox when checked values do not match options', () => {
    const { getByLabelText } = render(
      <FormCheckboxGroup
        value={['4', '5', '6']}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[
          { id: 'checkbox-1', label: 'Test checkbox 1', value: '1' },
          { id: 'checkbox-2', label: 'Test checkbox 2', value: '2' },
          { id: 'checkbox-3', label: 'Test checkbox 3', value: '3' },
        ]}
      />,
    );

    const selectAllCheckbox = getByLabelText('Select all') as HTMLInputElement;

    expect(selectAllCheckbox.checked).toBe(false);
  });

  test('does not render `Select all` checkbox when there is only one option', () => {
    const { queryByLabelText } = render(
      <FormCheckboxGroup
        value={[]}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        selectAll
        options={[{ id: 'checkbox-1', label: 'Test checkbox 1', value: '1' }]}
      />,
    );

    expect(queryByLabelText('Select all')).toBeNull();
    expect(queryByLabelText('Test checkbox 1')).not.toBeNull();
  });

  test('renders option with conditional contents', () => {
    const { container, getByText } = render(
      <FormCheckboxGroup
        value={['2']}
        id="test-checkboxes"
        name="test-checkboxes"
        legend="Test checkboxes"
        options={[
          {
            id: 'checkbox-1',
            label: 'Test checkbox 1',
            value: '1',
            conditional: <p>Conditional 1</p>,
          },
          {
            id: 'checkbox-2',
            label: 'Test checkbox 2',
            value: '2',
            conditional: <p>Conditional 2</p>,
          },
          {
            id: 'checkbox-3',
            label: 'Test checkbox 3',
            value: '3',
            conditional: <p>Conditional 3</p>,
          },
        ]}
      />,
    );

    const hiddenClass = 'govuk-checkboxes__conditional--hidden';

    expect(getByText('Conditional 1').parentElement).toHaveClass(hiddenClass);
    expect(getByText('Conditional 2').parentElement).not.toHaveClass(
      hiddenClass,
    );
    expect(getByText('Conditional 3').parentElement).toHaveClass(hiddenClass);
    expect(container.innerHTML).toMatchSnapshot();
  });
});
