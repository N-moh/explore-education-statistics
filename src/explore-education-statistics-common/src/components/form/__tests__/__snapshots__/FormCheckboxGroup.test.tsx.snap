// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormCheckboxGroup renders checkboxes with some pre-checked 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <div class="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-1"
               name="test-checkboxes"
               type="checkbox"
               value="1"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-1"
        >
          Test checkbox 1
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-2"
               name="test-checkboxes"
               type="checkbox"
               value="2"
               checked
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-2"
        >
          Test checkbox 2
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-3"
               name="test-checkboxes"
               type="checkbox"
               value="3"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-3"
        >
          Test checkbox 3
        </label>
      </div>
    </div>
  </fieldset>
</div>

`;

exports[`FormCheckboxGroup renders correctly with legend 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">
      Choose some checkboxes
    </legend>
    <div class="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-1"
               name="test-checkboxes"
               type="checkbox"
               value="1"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-1"
        >
          Test checkbox 1
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-2"
               name="test-checkboxes"
               type="checkbox"
               value="2"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-2"
        >
          Test checkbox 2
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-3"
               name="test-checkboxes"
               type="checkbox"
               value="3"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-3"
        >
          Test checkbox 3
        </label>
      </div>
    </div>
  </fieldset>
</div>

`;

exports[`FormCheckboxGroup renders correctly with small size variants 1`] = `
<div>
  <div
    class="govuk-form-group"
  >
    <fieldset
      class="govuk-fieldset"
      id="test-checkboxes"
    >
      <div
        class="govuk-checkboxes"
      >
        <div
          class="govuk-checkboxes__item itemSmall"
        >
          <input
            class="govuk-checkboxes__input"
            id="checkbox-1"
            name="test-checkboxes"
            type="checkbox"
            value="1"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="checkbox-1"
          >
            Test checkbox 1
          </label>
        </div>
        <div
          class="govuk-checkboxes__item itemSmall"
        >
          <input
            class="govuk-checkboxes__input"
            id="checkbox-2"
            name="test-checkboxes"
            type="checkbox"
            value="2"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="checkbox-2"
          >
            Test checkbox 2
          </label>
        </div>
        <div
          class="govuk-checkboxes__item itemSmall"
        >
          <input
            class="govuk-checkboxes__input"
            id="checkbox-3"
            name="test-checkboxes"
            type="checkbox"
            value="3"
          />
          <label
            class="govuk-label govuk-checkboxes__label"
            for="checkbox-3"
          >
            Test checkbox 3
          </label>
        </div>
      </div>
    </fieldset>
  </div>
</div>
`;

exports[`FormCheckboxGroup renders list of checkboxes in correct order 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <div class="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-1"
               name="test-checkboxes"
               type="checkbox"
               value="1"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-1"
        >
          Test checkbox 1
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-2"
               name="test-checkboxes"
               type="checkbox"
               value="2"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-2"
        >
          Test checkbox 2
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-3"
               name="test-checkboxes"
               type="checkbox"
               value="3"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-3"
        >
          Test checkbox 3
        </label>
      </div>
    </div>
  </fieldset>
</div>

`;

exports[`FormCheckboxGroup renders unchecked \`Select all\` option when \`selectAll\` is true 1`] = `

<div class="govuk-form-group">
  <fieldset class="govuk-fieldset"
            id="test-checkboxes"
  >
    <div class="govuk-checkboxes">
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="test-checkboxes-all"
               name="test-checkboxes"
               type="checkbox"
               value="select-all"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="test-checkboxes-all"
        >
          Select all
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-1"
               name="test-checkboxes"
               type="checkbox"
               value="1"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-1"
        >
          Test checkbox 1
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-2"
               name="test-checkboxes"
               type="checkbox"
               value="2"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-2"
        >
          Test checkbox 2
        </label>
      </div>
      <div class="govuk-checkboxes__item">
        <input class="govuk-checkboxes__input"
               id="checkbox-3"
               name="test-checkboxes"
               type="checkbox"
               value="3"
        >
        <label class="govuk-label govuk-checkboxes__label"
               for="checkbox-3"
        >
          Test checkbox 3
        </label>
      </div>
    </div>
  </fieldset>
</div>

`;
